refactoring/clean up module transformers

cos opened the pull request:
Main changes:
- removed the use of "hybrid" from all but the front-end code
- most transformers now inherit from ModuleTransformer, which takes care of correct recursive application

cos commented:
@daejunpark, please review

cos commented:
@daejunpark is at OOPSLA. anyone else, please feel free to pick this up.

radumereuta commented on k-distribution/src/test/java/org/kframework/kore/convertors/BubbleParsing.java:
```diff
@@ -95,8 +97,8 @@ public BubbleParsing(URL koreSyntax, String mainModule, String startSort) {
      * TODO(radu): generalize this function, and eliminate duplicates
```
Is this comment still valid?

cos replied:
no, it looks like you already made it nicer at some point.

one thing that could be done here is have the class implement `SentenceBasedModuleTransformer` instead of `BasicModuleTransformer`. The body of the map would become the implementation of `f` and the rest of the code in this function would go away.

radumereuta commented on kernel/src/main/java/org/kframework/kore/compile/ResolveAnonVar.java:
```diff
@@ -43,7 +44,7 @@ private Context resolve(Context context) {
     }
 
-    public synchronized Sentence resolve(Sentence s) {
+    public synchronized Sentence f(Sentence s) {
```
This should have a better name. At least JavaDoc.

radumereuta commented on kernel/src/main/java/org/kframework/kore/compile/ResolveHeatCoolAttribute.java:
```diff
@@ -55,7 +56,7 @@ private K transform(K requires, Att att) {
     }
 
-    public Sentence resolve(Sentence s) {
+    public Sentence f(Sentence s) {
```
Again, better name/JavaDoc, since it's public.

radumereuta commented on kernel/src/main/java/org/kframework/kore/compile/ResolveSemanticCasts.java:
```diff
@@ -124,7 +125,7 @@ public K apply(KVariable k) {
 
 
-    public synchronized Sentence resolve(Sentence s) {
+    public synchronized Sentence f(Sentence s) {
```
dito

radumereuta commented on kore/src/main/scala/org/kframework/definition/transformers.scala:
```diff
@@ -142,14 +120,38 @@ trait HybridModuleTransformer extends ModuleTransformer {
 }
 
-abstract class WithInputDefinitionModuleTransformer(inputDefinition: Definition) extends BasicModuleTransformer {
+abstract class SentenceBasedModuleTransformer extends BasicModuleTransformer {
```
Ah, this is where they come from.
This should have some JavaDoc. Maybe you can skip the previous ones if this class becomes easy to understand.

cos commented:
@radumereuta, I addressed all your comments.

The only place with remaining "hybrid" transformers is the frontend. Would you have time to have a go at removing them from there as well at some point? Or we could work together on it if you like.

radumereuta commented:
lgtm
